New Async model & Allow space in repos or users search

taooceros opened the pull request:
1. Use AsyncPlugin to allow cancellation token & real async execution
2. Allow space in repos or users search (refactor pattern matching)

taooceros commented:
Rider does the large format change. Hope that's not too bad 😂.

I am not familiar with F# before contacting this project, so hope the code is not too bad.

JohnTheGr8 commented:
Can you please turn off Rider's formatting and re-submit (or force-push) just the actual changes? Sorry to trouble you but the formatting changes are just too much in this case, it's hard to even spot the non-formatting changes... 😂

taooceros commented:
Sure🤣

taooceros committed:
New Async model & Allow space in repos or users search
1. Use AsyncPlugin to allow cancellation token & real async execution
2. Allow space in repos or users search (refactor pattern matching)

taooceros commented:
Done

JohnTheGr8 commented:
Thanks @taooceros that's better!

I understand that this is your first time dealing with F#, hopefully it was a fun experience so far.. I know there was a lot of head-scratching for me at first 😆

Is it okay if I guide you through a few changes? Some changes you introduced could be achieved in simpler ways. Also, you probably didn't notice but the tests don't compile now because of the signature change of the `ProcessQuery` member.

taooceros commented:
> I understand that this is your first time dealing with F#, hopefully it was a fun experience so far.. I know there was a lot of head-scratching for me at first 😆

> Is it okay if I guide you through a few changes? Some changes you introduced could be achieved in simpler ways. Also, you probably didn't notice but the tests don't compile now because of the signature change of the ProcessQuery member.

Sure! Thank you! Learning F# is quite interesting for me lol!
I will fix the test soon.

JohnTheGr8 commented on src/GithubPlugin.fs:
```diff
@@ -127,27 +136,42 @@ type GithubPlugin() =
-        >> function
-            | Choice1Of2 result -> presentApiSearchResult result
-            | Choice2Of2 exn -> presentApiSearchExn exn
+    let tryRunApiSearch (result: Async<ApiSearchResult>) =
+        async {
+            let! res = result |> Async.Catch
 
+            return
+                match res with
+                | Choice1Of2 result -> presentApiSearchResult result
+                | Choice2Of2 exn -> presentApiSearchExn exn
+        }
```
This can be simplified to

```fsharp
let tryRunApiSearch =
  Async.Catch
  >> Async.map (function
        | Choice1Of2 result -> presentApiSearchResult result
        | Choice2Of2 exn -> presentApiSearchExn exn)
```

taooceros replied:
hmm why rider suggests that we don't have `Async.map` as a function?

JohnTheGr8 replied:
Ah, my bad, I thought `Async.map` was included in core...

In that case you can simplify it down to something like:

```fsharp
let tryRunApiSearch fSearch =
    async {
        match! Async.Catch fSearch with
        | Choice1Of2 result -> return presentApiSearchResult result
        | Choice2Of2 exn -> return presentApiSearchExn exn
    }
```

JohnTheGr8 commented:
> Sure! Thank you! Learning F# is quite interesting for me lol!
> I will fix the test soon.

Hold on, let me write down a few changes before you change anything else.

JohnTheGr8 commented on src/GithubPlugin.fs:
```diff
@@ -18,21 +20,28 @@ type GithubPlugin() =
-        | [ UserReposFormat user ]                -> runApiSearch (FindUserRepos user)
-        | [ search ]                              -> SuggestQuery (SearchRepos search)
-        | _                                       -> SuggestQuery DefaultSuggestion
+    let (|SpecificSearchFormat|_|) keyword =
+        match keyword with
+        | "issues" -> Some(FindIssues)
+        | "pr"
+        | "pull" -> Some(FindPRs)
+        | "repo" -> Some(FindRepo)
+        | _ -> None
+
+    let (|GeneralSearchFormat|_|) keyword =
+        match keyword with
+        | "repos" -> Some(FindRepos)
+        | "users" -> Some(FindUsers)
+        | _ -> None
+
+    let parseQuery (searchTerm: Query) =
+        match (searchTerm.FirstSearch, searchTerm.SecondToEndSearch) with
+        | GeneralSearchFormat keywordFunc, search when search.Length > 0 -> search |> keywordFunc |> runApiSearch
+        | SpecificSearchFormat keywordFunc, UserRepoFormat search
+        | UserRepoFormat search, SpecificSearchFormat keywordFunc -> search |> keywordFunc |> runApiSearch
+        | UserRepoFormat (u, r), IssueFormat i -> runApiSearch (FindIssue(u, r, i))
+        | search, "" -> SuggestQuery(SearchRepos search)
+        | _ -> SuggestQuery DefaultSuggestion
```
Here you could have kept the `parseQuery` function as-is except for the first two match cases. 

On the matching side, you can use the `::` operator for these two cases to deconstruct the input into the head (either "repos" or "users") and the tail (list of space separated query words).

On the right side, you can re-join the tail into a single string `String.concat " "`

taooceros replied:
I check Flow's code. It may be better to let Flow handle the split because the Term field is simply split from user search, so if the actionkeyword is *, it would be possible to skip a search term unexpected.

JohnTheGr8 replied:
Good point, then perhaps something like

```fsharp
|> List.skip (if query.ActionKeyword = query.GlobalPluginWildcardSign then 0 else 1)
```

in `QueryAsync` ?

JohnTheGr8 replied:
(though, realistically, who in their right mind would use * for the github plugin 😛)

taooceros replied:
I extract the list from Query.FirstSearch and Query.SecondToEndSearch. What do you think about that👀

JohnTheGr8 replied:
I prefer the old way for a few reasons...

The way you have it, `parseQuery` will always get a list of exactly 2 elements, which is misleading and defeats the purpose of passing a list, and also does not allow for pattern matching to be fully utilised on the typed query. You also have to do awkward changes to the tests like you did in 352e26a4caf7e1629dd18d4e960519c69a19dcec to get them passing...

I also think it's much cleaner to have all query cases in a single function instead of splitting them up, I find them easier to read/reason about that way.

Another minor point is that I'd like to keep the differences between Wox and Flow plugins to the minimum.

Anyway, I haven't tested the following but I'm thinking of something like this:

```fsharp
let parseQuery = function
    | "repos" :: search                       -> runApiSearch (FindRepos (String.concat " " search))
    | "users" :: search                       -> runApiSearch (FindUsers (String.concat " " search))
    | [ "issues"; UserRepoFormat search ]     -> runApiSearch (FindIssues search)
    | [ "pr";     UserRepoFormat search ]     -> runApiSearch (FindPRs search)
    | [ "pull";   UserRepoFormat search ]     -> runApiSearch (FindPRs search)
    | [ "repo";   UserRepoFormat search ]     -> runApiSearch (FindRepo search)
    | [ UserRepoFormat search           ]     -> runApiSearch (FindRepo search)
    | [ UserRepoFormat search; "issues" ]     -> runApiSearch (FindIssues search)
    | [ UserRepoFormat search; "pr"     ]     -> runApiSearch (FindPRs search)
    | [ UserRepoFormat search; "pull"   ]     -> runApiSearch (FindPRs search)
    | [ UserRepoFormat (u,r); IssueFormat i ] -> runApiSearch (FindIssue (u,r,i))
    | [ UserReposFormat user ]                -> runApiSearch (FindUserRepos user)
    | [ search ]                              -> SuggestQuery (SearchRepos search)
    | _                                       -> SuggestQuery DefaultSuggestion
```

and then the `QueryAsync` overload can look like this:

```fsharp
member this.QueryAsync(query, token: CancellationToken) =
    let ghSearch = async {
        let! results =
            query.Terms
            |> List.ofArray
            |> List.skip (if query.ActionKeyword = query.GlobalPluginWildcardSign then 0 else 1)
            |> this.ProcessQuery

        return results
                |> List.map (fun r -> Result( Title = r.title, SubTitle = r.subtitle, IcoPath = "icon.png", Action = fun x -> r.action x ))
                |> List<Result>
    }
    Async.StartImmediateAsTask(ghSearch, token)
```

taooceros replied:
Got you😉

JohnTheGr8 commented on src/GithubPlugin.fs:
```diff
@@ -127,27 +136,42 @@ type GithubPlugin() =
-
-    interface IPlugin with
-        member this.Init (context:PluginInitContext) =
+        async {
+            let! res =
+                match parseQuery terms with
+                | RunApiSearch fSearch -> tryRunApiSearch fSearch
+                | SuggestQuery suggestion -> async { return presentSuggestion suggestion }
+
+            return
+                res
+                |> List.map
+                    (fun r ->
+                        Result(
+                            Title = r.title,
+                            SubTitle = r.subtitle,
+                            IcoPath = "icon.png",
+                            Action = fun x -> r.action x
+                        ))
+                |> List<Result>
+        }
```
`ProcessQuery` is meant as an intermediate function here, used in testing, which is why it should return F#-friendly types (`SearchResult list` instead of `List<Flow.Launcher.Plugin.Result>`)

Moving the mapping to `List<Flow.Launcher.Plugin.Result>` out of `ProcessQuery` will get you most of the way there to getting the tests back. 

The remaining issue to resolve is that `ProcessQuery` is now wrapped in `Async`. To resolve this, you can add a helper function in the tests code that wraps `ProcessQuery` with `Async.RunSynchronously` and then replace all instances of `plugin.ProcessQuery` with it.

taooceros replied:
I check the test framework document. They have a testAsync CE, so we can simply replace test with testAsync😉

JohnTheGr8 replied:
Yeah I know, I use it daily 😛 I just suggested the more straightforward change to keep things simple.

JohnTheGr8 commented on src/GithubPlugin.fs:
```diff
@@ -127,27 +136,42 @@ type GithubPlugin() =
-            |> this.ProcessQuery
-            |> List.map (fun r -> Result( Title = r.title, SubTitle = r.subtitle, IcoPath = "icon.png", Action = fun x -> r.action x ))
-            |> List<Result>
+        member this.QueryAsync(query, token: CancellationToken) =
+            Async.StartImmediateAsTask(this.ProcessQuery query, token)
```
Continuing from above, you can declare a local async function here that passes `query.Terms` down to `ProcessQuery` and then maps the results to a `List<Flow.Launcher.Plugin.Result>`, and then pass that function to `Async.StartImmediateAsTask`.

JohnTheGr8 replied:
So to clarify, the local function could look something like:

```fsharp
let f = async {
    let! results = 
        query.Terms
        |> List.ofArray
        |> List.skip 1
        |> this.ProcessQuery
    
    return results
           |> List.map (fun r -> Result( Title = r.title, SubTitle = r.subtitle, IcoPath = "icon.png", Action = fun x -> r.action x ))
           |> List<Result>
}
```

JohnTheGr8 reviewed (changes requested):
I've tried to explain at some length but if something doesn't make sense let me know and I'll try harder ;)

taooceros committed:
Refactor Main Search Code

Co-Authored-By: Ioannis G. <jtgftw@gmail.com>

taooceros committed:
Move test to Async

taooceros committed:
Rename & Add testing project to sln

taooceros committed:
Fix testing

JohnTheGr8 commented on tests/Main.fs:
```diff
@@ -26,64 +26,74 @@ let allTests =
-            plugin.ProcessQuery [ "repos"; "wox" ] |> Expect.isNonEmpty "should not be empty"
+            let! res = plugin.ProcessQuery [ "repos"; "wox" ]
+            
+            res|>
+            Expect.isNonEmpty "should not be empty"
```
nitpick: make this one line and add space after `res`

JohnTheGr8 commented on tests/Main.fs:
```diff
@@ -26,64 +26,74 @@ let allTests =
-            plugin.ProcessQuery [ "users"; "john" ] |> Expect.isNonEmpty "should not be empty"
+            let! res = plugin.ProcessQuery [ "users"; "john" ]
+            
+            res |>
+            Expect.isNonEmpty "should not be empty"
```
nitpick: make this one line as well

taooceros committed:
Adjust Pattern Matching

Co-Authored-By: Ioannis G. <jtgftw@gmail.com>

taooceros committed:
await 300ms for actual api search

JohnTheGr8 committed:
add test case for repo search with spaces

JohnTheGr8 commented:
hey @taooceros - sorry for putting this on hold for a bit...

Presumably the 300ms delay you introduced in the last commit is to avoid starting API searches while the user is typing (because Octokit doesn't support passing a cancellation token) ?

JohnTheGr8 committed:
quick refactoring

taooceros commented:
No worries, yes

JohnTheGr8 commented:
Did a manual squash & merge, github doesn't pick that up 🤷‍♂️ 

Thanks @taooceros
